### DISCUSSION TITLE
Index Initializers for Record 'with' expressions

### DISCUSSION THREAD

**Discussion by altworowsky** [dotnet/csharplang#8018] [Language Ideas]
In C# 12, we're able to define our own object initializers with indexers or 'index initializers' for records.
```csharp
var m = new Matrix {
  [0, 0] = 1.0,
  [1, 1] = 2.0,
  [2, 2] = 4.0
};
```

But, we're not able to use those index initializers with a 'with' expression.
```csharp
var n = m with {
  [8, 8] = 64.0,
  [10, 10] = 1024.0
};
```

The latter feels like a natural interaction between the index initializer and 'with' expression syntaxes. I would suggest that the language be amended to allow for the use of indexer initializers with 'with' expressions.

**Comment by HaloFour** [comment]
That could be a bit awkward.  Record cloning is shallow, so any internal collection state would be shared between the two instances and you would end up mutating the first instance.

**Reply by altworowsky** [reply]
It's possible to define custom cloning semantics to avoid this issue.[^1]

[^1]: [Custom Copy Semantics](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/with-expression#custom-copy-semantics)

**Reply by HaloFour** [reply]
Yes, you can define a custom copy constructor, but the onus would be on the developer to know to do that and to know how to make a clone of the collection.

I can't seem to find any mentions of other initializers being supported with "wither" expressions so I don't know if this case was considered.

**Reply by altworowsky** [reply]
That's a general problem when designing classes that can be cloned or copied with internally referenced objects. I don't believe it's a unique problem created or compounded by the suggested syntax. Developers already need to consider the problem if their record has reference type fields and could be cloned with a 'with' expression or otherwise.

**Reply by HaloFour** [reply]
Indeed, which is a core feature of records.  That's why it's probably not wise to expose an indexer initializer, or collection initializers, on record withers, by default you'd be doing it wrong.

I'm curious if it came up.  The meeting notes don't seem to mention it.

**Reply by altworowsky** [reply]
By that line of reasoning, we wouldn't allow reference types in a record because you'd be doing it wrong by default.

We already ask developers to take care with reference types in records. The proposed syntax doesn't ask for any additional care that's not already asked for.

I couldn't find any prior discussions either. Also, thanks for your engagement on the suggestion.
